### PR TITLE
Client/server worker fixes

### PR DESCRIPTION
**Pull request by vietj**
@purplefox please review

### CODE REVIEW

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java [hunk 1]
```diff
@@ -146,6 +149,7 @@ synchronized void toWebSocket(String requestURI, MultiMap headers, WebsocketVers
       p.addBefore("handler", "handshakeCompleter", new HandshakeInboundHandler(wsConnect, version != WebSocketVersion.V00));
       handshaker.handshake(channel).addListener(future -> {
         if (!future.isSuccess() && exceptionHandler != null) {
+          // On good thread ?
```
Comment (purplefox) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: ?

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java [hunk 2]
```diff
@@ -183,46 +187,48 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
 
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
-      context.executeFromIO(() -> {
-        synchronized (ClientConnection.this) {
-          if (handshaker != null && handshaking) {
-            if (msg instanceof HttpResponse) {
-              HttpResponse resp = (HttpResponse) msg;
-              if (resp.getStatus().code() != 101) {
+      synchronized (ClientConnection.this) {
```
Comment (purplefox) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: This doesn't look right. The synchronized block needs to be inside the context.executeIO - this is to ensure that there is a memory barrier over access to ClientConnection internal state.

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java [hunk 3]
```diff
@@ -242,8 +248,12 @@ private void handshakeComplete(ChannelHandlerContext ctx, FullHttpResponse respo
       }
       ws = new WebSocketImpl(vertx, ClientConnection.this, supportsContinuation, client.getOptions().getMaxWebsocketFrameSize());
       handshaker.finishHandshake(channel, response);
+
       log.debug("WebSocket handshake complete");
-      wsConnect.handle(ws);
+      context.executeFromIO(() -> {
```
Comment (purplefox) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: handshake complete is executed from channelread which is already executed on the right thread so this appears to be doing it twice.

Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: when I execute the test (WebsocketTest#testClientWorker), the channel thread is an eventloop thread, not sure to understand what you mean.

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java (hunk 2 above)
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: At this point the thread is an event loop thread. So it means we need it in the executeFromIO smaller blocks.

**Pull request closed without merging**